reconcile/external_resources/model.py

```diff
@@ -0,0 +1,221 @@
+import base64
+import hashlib
+import json
+from abc import (
+    ABC,
+)
+from collections.abc import ItemsView, Iterable, KeysView, ValuesView
+from enum import Enum
+from typing import Any
+
+from pydantic import BaseModel
+
+from reconcile.gql_definitions.external_resources.external_resources_modules import (
+    ExternalResourcesModuleV1,
+)
+from reconcile.gql_definitions.external_resources.external_resources_namespaces import (
+    NamespaceTerraformProviderResourceAWSV1,
+    NamespaceTerraformResourceRDSV1,
+    NamespaceTerraformResourceRoleV1,
+    NamespaceV1,
+)
+from reconcile.utils.external_resource_spec import (
+    ExternalResourceSpec,
+)
+
+
+class ExternalResourceValidationError(Exception):
+    errors: list[str] = []
+
+    def add_validation_error(self, msg: str) -> None:
+        self.errors.append(msg)
+
+
+class ExternalResourceKey(BaseModel, frozen=True):
+    provision_provider: str
+    provisioner_name: str
+    provider: str
+    identifier: str
+
+    @staticmethod
+    def from_spec(spec: ExternalResourceSpec) -> "ExternalResourceKey":
+        return ExternalResourceKey(
+            provision_provider=spec.provision_provider,
+            provisioner_name=spec.provisioner_name,
+            identifier=spec.identifier,
+            provider=spec.provider,
+        )
+
+    def hash(self) -> str:
+        return hashlib.md5(
+            json.dumps(self.dict(), sort_keys=True).encode("utf-8")
+        ).hexdigest()
+
+    @property
+    def state_path(self) -> str:
+        return f"{self.provision_provider}/{self.provisioner_name}/{self.provider}/{self.identifier}"
+
+
+class ExternalResourcesInventory:
```

Comment:
Looks like a Mapping, perhaps inherit from MutatableMapping class ?

Reply:
Yup, that's right. I moved this convenience method to the `ExternalResourcesInventory`class. PTAL.